### DISCUSSION TITLE
Learning weights of a network for mimicking biological neural network activity

### DISCUSSION THREAD

**Discussion by fire-papaya** [jeshraghian/snntorch#137] [Q&A]
I am trying to learn weights of a simple feed forward network based on results of a simulation.
The setup is: 50-channels spike train input, 50-neuron hidden layer, 8-neuron output layer. The simulation is performed over 1sec with resolution of 0.1 ms, so 10_000 datapoints. My task is: given network **A**, its input and output, learn its weights in order to mimic its activity in network **B**. 

Currently I have problems with defining a proper loss criterion, since the output of A is not classification, but just a spike train of several neurons. Here is the draft code for a single forward-prop and loss computation

```
import snntorch as snn
from snntorch import functional as SF
from snntorch import utils
from snntorch import spikeplot as splt

import torch
import torch.nn as nn

from helpers import read_spikes_from_file


def forward_pass(net, num_steps, data):
    mem_rec = []
    spk_rec = []
    utils.reset(net)  # resets hidden states for all LIF neurons in net

    for step in range(num_steps):
        spk_out, syn_out, mem_out = net(data[step])
        spk_rec.append(spk_out)
        mem_rec.append(mem_out)

    return torch.stack(spk_rec), torch.stack(mem_rec)


duration = 10000
# - Define the network size (for now alpha and beta are predefined)
input_size = 50
hidden_size = 50
output_size = 8
beta = 0.99
alpha = 0.9

input_spikes = read_spikes_from_file('pop1_spikes.dat', shape=(duration, input_size)).unsqueeze(1)

target_spikes = read_spikes_from_file('output_spikes.dat', shape=(duration, output_size)).unsqueeze(1)
# simulation used Conductance-based LIF neuron, so I used Synaptic model here
network = nn.Sequential(
    nn.Linear(input_size, hidden_size),
    snn.Synaptic(alpha=alpha, beta=beta, init_hidden=True),
    nn.Linear(hidden_size, output_size),
    snn.Synaptic(alpha=alpha, beta=beta, init_hidden=True, output=True)
)

output_spikes, _ = forward_pass(network, duration, input_spikes)

loss_fn = SF.ce_rate_loss()

loss_val = loss_fn(output_spikes, target_spikes)

print(f"The loss from an untrained network is {loss_val.item():.3f}")
```

**Comment by ahenkes1** [comment]
Have you tried MSE? And if T is magnitudes greater than the feature vector, you might try recurrent networks, as it can be difficult to sustain the information over lot's of time steps.

And why do you reset the hidden states for every forward pass?
